src/hotspot/share/gc/g1/g1ConcurrentRefine.cpp

```diff
@@ -345,86 +200,192 @@ uint G1ConcurrentRefine::max_num_threads() {
   return G1ConcRefinementThreads;
 }
 
-static size_t calc_new_green_zone(size_t green,
-                                  double logged_cards_scan_time,
-                                  size_t processed_logged_cards,
-                                  double goal_ms) {
-  // Adjust green zone based on whether we're meeting the time goal.
-  // Limit to max_green_zone.
-  const double inc_k = 1.1, dec_k = 0.9;
-  if (logged_cards_scan_time > goal_ms) {
-    if (green > 0) {
-      green = static_cast<size_t>(green * dec_k);
+void G1ConcurrentRefine::update_pending_cards_target(double logged_cards_scan_time_ms,
+                                                     size_t processed_logged_cards,
+                                                     size_t predicted_thread_buffer_cards,
+                                                     double goal_ms) {
+  size_t minimum = minimum_pending_cards_target();
+  if ((processed_logged_cards < minimum) || (logged_cards_scan_time_ms == 0.0)) {
+    log_debug(gc, ergo, refine)("Unchanged pending cards target: %zu",
+                                _pending_cards_target);
+    return;
+  }
+
+  // Base the pending cards budget on the measured rate.
+  double rate = processed_logged_cards / logged_cards_scan_time_ms;
+  size_t budget = static_cast<size_t>(goal_ms * rate);
+  // Deduct predicted cards in thread buffers to get target.
+  size_t new_target = budget - MIN2(budget, predicted_thread_buffer_cards);
+  // Add some hysterisis with previous values.
+  if (is_pending_cards_target_initialized()) {
+    new_target = (new_target + _pending_cards_target) / 2;
+  }
+  // Apply minimum target.
+  new_target = MAX2(new_target, minimum_pending_cards_target());
+  _pending_cards_target = new_target;
+  log_debug(gc, ergo, refine)("New pending cards target: %zu", new_target);
+}
+
+void G1ConcurrentRefine::adjust_after_gc(double logged_cards_scan_time_ms,
+                                         size_t processed_logged_cards,
+                                         size_t predicted_thread_buffer_cards,
+                                         double goal_ms) {
+  if (!G1UseConcRefinement) return;
+  update_pending_cards_target(logged_cards_scan_time_ms,
+                              processed_logged_cards,
+                              predicted_thread_buffer_cards,
+                              goal_ms);
+  if (_thread_control.max_num_threads() == 0) {
+    // If no refinement threads then the mutator threshold is the target.
+    _dcqs.set_max_cards(_pending_cards_target);
+  } else {
+    // Provisionally make the mutator threshold unlimited, to be updated by
+    // the next periodic adjustment.  Because card state may have changed
+    // drastically, record that adjustment is needed and kick the primary
+    // thread, in case it is waiting.
+    _dcqs.set_max_cards(SIZE_MAX);
```

Comment:
Don't we already set `_dcqs.set_max_cards(SIZE_MAX)` in `G1DirtyCardQueueSet::concatenate_logs()`

Reply:
We do, but I don't want to rely on that far away and done for other reasons
code. The reason for setting in concatenate_logs is to cut off any refinement
by handle_completed_buffer (via flushing buffers). It isn't entirely necessary
for that either, as there are other reasons why that function won't do any
refinement. I waffled about whether to keep or remove it.